### PR TITLE
perf: Improve parser performance for typescript

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

The ts parsing time is reduced by ~20%, and the general js parsing time is increased by ~2%.
If we want general js not to be affected, maybe we can manually replace all `getters/setters`.

```
PS F:\babel\benchmark\babel-parser\real-case-ts> node .\bench.mjs
current 1 typescript parser.ts: 67.9 ops/sec ±1.01% (15ms)
baseline 1 typescript parser.ts: 55.09 ops/sec ±1.22% (18ms)
current 1 typescript parser.ts: 68.58 ops/sec ±0.79% (15ms)
baseline 1 typescript parser.ts: 55.88 ops/sec ±0.84% (18ms)
PS F:\babel\benchmark\babel-parser\real-case-ts> node .\bench.mjs
current 1 typescript parser.ts: 66.91 ops/sec ±0.98% (15ms)
baseline 1 typescript parser.ts: 53.97 ops/sec ±1.14% (19ms)
current 1 typescript parser.ts: 68.01 ops/sec ±0.75% (15ms)
baseline 1 typescript parser.ts: 55.14 ops/sec ±0.81% (18ms)
```
```
PS F:\babel\benchmark\babel-parser\real-case> node .\bench.mjs
current 4 jquery 3.6: 24.43 ops/sec ±1.95% (41ms)
baseline 4 jquery 3.6: 25.48 ops/sec ±1.11% (39ms)
current 4 jquery 3.6: 25.01 ops/sec ±1.7% (40ms)
baseline 4 jquery 3.6: 25.2 ops/sec ±1.35% (40ms)
PS F:\babel\benchmark\babel-parser\real-case> node .\bench.mjs
current 4 jquery 3.6: 24.46 ops/sec ±1.78% (41ms)
baseline 4 jquery 3.6: 24.89 ops/sec ±1.46% (40ms)
current 4 jquery 3.6: 24.61 ops/sec ±1.67% (41ms)
baseline 4 jquery 3.6: 25.16 ops/sec ±1.26% (40ms)
```

### CODE REVIEW

**Comment by liuxingbaoyu** [comment]
[`a25c777` (#16072)](https://github.com/babel/babel/pull/16072/commits/a25c777e7f33b931f52c82567efe4c9c916099ac)
This commit will only have a slight performance impact on errors and will not affect overall performance, but it simplifies our code.

File: packages/babel-parser/src/tokenizer/state.ts [hunk 1]
```diff
@@ -152,20 +279,36 @@ export default class State {
   }
 
   clone(skipArrays?: boolean): State {
```
Comment (liuxingbaoyu) [inline packages/babel-parser/src/tokenizer/state.ts:281]: Actually we don't use `skipArrays` anywhere. :)

File: packages/babel-parser/src/tokenizer/state.ts [hunk 2]
```diff
@@ -152,20 +279,36 @@ export default class State {
...
+    state.type = this.type;
+    state.value = this.value;
+    state.start = this.start;
+    state.end = this.end;
+    state.lastTokEndLoc = this.lastTokEndLoc;
+    state.lastTokStartLoc = this.lastTokStartLoc;
+    state.context = maybeCopy(this.context);
+    state.firstInvalidTemplateEscapePos = this.firstInvalidTemplateEscapePos;
+    state.strictErrors = this.strictErrors;
+    state.tokensLength = this.tokensLength;
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/tokenizer/state.ts:310]: Do you think we can somehow type-check this list to make sure we don't forget to add props? If not, can you add a comment at the beginning of `State`'s body saying that when adding a new property we must also add it here?

File: packages/babel-parser/src/parser/comments.ts [hunk 3]
```diff
@@ -102,7 +102,10 @@ function adjustInnerComments(
 export default class CommentsParser extends BaseParser {
   addComment(comment: Comment): void {
     if (this.filename) comment.loc.filename = this.filename;
-    this.state.comments.push(comment);
+    const { commentsLen } = this.state;
+    if (this.comments.length != commentsLen) this.comments.length = commentsLen;
+    this.comments.push(comment);
+    this.state.commentsLen++;
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/comments.ts:108]: Can it happen that we clone the state, push a comment to `.comments`, reset to the previous state, then never call `.addComment` again and so we never do `this.comments.length = commentsLen` removing the new extra comment?

**Review by nicolo-ribaudo — COMMENTED** [review]
I think given how popular TS is, the perf change is ok.

File: packages/babel-parser/src/tokenizer/state.ts (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-parser/src/tokenizer/state.ts:286]: Can we just remove it then? :)

File: packages/babel-parser/src/util/scope.ts [hunk 4]
```diff
@@ -155,42 +166,39 @@ export default class ScopeHandler<IScope extends Scope = Scope> {
...
   checkLocalExport(id: N.Identifier) {
     const { name } = id;
     const topLevelScope = this.scopeStack[0];
     if (
-      !topLevelScope.lexical.has(name) &&
-      !topLevelScope.var.has(name) &&
+      !topLevelScope.names.has(name)
       // In strict mode, scope.functions will always be empty.
       // Modules are strict by default, but the `scriptMode` option
       // can overwrite this behavior.
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/util/scope.ts:201]: This comments can be removed, right?

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]
Change summary for future reviewers:
1. Pack tokenizer boolean properties into bit flags
1. The tokenizer state access `.lastTokStart` is replaced by `.lastTokStartLoc.index`
1. Replace `.raise({ at })` by `.raise(at)` to reduce cost of intermediate object literal
1. Specify each state properties in parser state clone to avoid dynamic property access
1. Merge multiple names set in parser scope tracker into a map

**Pull request merged**